src/main/java/io/cdap/plugin/http/source/common/pagination/BaseHttpPaginationIterator.java

```diff
@@ -15,15 +15,182 @@
+    }
+
+    // response being null, means it's the first page we are loading
+    if (response != null && config.getWaitTimeBetweenPages() != null) {
+      Thread.sleep(config.getWaitTimeBetweenPages());
+    }
+
+    LOG.debug("Fetching '{}'", nextPageUrl);
+
+    PollInterval pollInterval;
+    if (config.getRetryPolicy().equals(RetryPolicy.LINEAR)) {
+      pollInterval = FixedPollInterval.fixed(config.getLinearRetryInterval(), TimeUnit.SECONDS);
+    } else {
+      pollInterval = IterativePollInterval.iterative(duration -> duration.multiply(2));
+    }
+
+    try {
+      Awaitility
+        .await().with()
+        .pollInterval(pollInterval)
+        .timeout(config.getMaxRetryDuration(), TimeUnit.SECONDS)
+        .until(this::visitPageAndCheckStatusCode);
+    } catch (ConditionTimeoutException ex) {
+      // Retries failed. We don't need to do anything here. This will ba handled using httpStatusCode below.
+    }
+
+    HttpEntity reponseEntity = response.getEntity();
+    byte[] responseBytes = EntityUtils.toByteArray(reponseEntity);
+    if (config.getFormat().equals(PageFormat.BLOB)) {
+      responseBytes = Base64.encodeBase64(responseBytes);
+    }
+    String responseBody = new String(responseBytes, StandardCharsets.UTF_8);
+
+    HttpErrorHandlingStrategy postRetryStrategy = httpErrorHandler.getErrorHandlingStrategy(httpStatusCode)
+      .getAfterRetryStrategy();
+
+    String errorMessage = String.format("Fetching from url '%s' returned status code '%d' and body '%s'",
+                                        nextPageUrl, httpStatusCode, responseBody);
+
+    switch (postRetryStrategy) {
+      case SUCCESS:
+        break;
+      case FAIL:
+        throw new IllegalStateException(errorMessage);
+      case SKIP:
+      case SEND_TO_ERROR:
+        if (!this.supportsSkippingPages()) {
+          throw new IllegalStateException(String.format(
+            "Pagination type '%s', does not support 'skip' and 'send to error' error handling.",
+            config.getPaginationType()));
+        }
+        LOG.warn(errorMessage);
+        // this will be handled by PageFactory. Here no handling is needed.
+        break;
+      default:
+        throw new IllegalArgumentException(String.format("Unexpected http error handling: '%s'", postRetryStrategy));
+    }
+
+    BasePage page = PageFactory.createInstance(config, responseBody, postRetryStrategy);
+    nextPageUrl = getNextPageUrl(responseBody, response, page);
+
+    LOG.debug("Next Page Url is '{}'", nextPageUrl);
+    //LOG.info("Schema is: {}", config.getSchemaString());
+
+    return page;
+  }
+
+  /**
+   * @return true if page still has elements to iterate. Otherwise it will load next page.
+   * False if no more pages to load or the page loaded has no elements.
+   */
+  protected boolean ensurePageIterable() {
+    try {
+      if (page == null || !page.hasNext()) {
+        page = getNextPage();
+      }
+
+      return (page != null && page.hasNext());
```

Comment:
Page instance will never be null. see PageFactory#createInstance it cannot return null.

Reply:
done

Reply:
Empty page meaning one where page.hasNext() is false.

If page.hasNext() is false a getNextPage() would have returned a non-null, then this will terminate early.

Reply:
Empty page (with no elements) in pagination APIs usually means that you should stop paginating otherwise you'll get into endless loop.

Example API:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=80

You can go to page 100 200 500 1000 which are empty and it's endless. All with no elements.
Page 10.000:
https://services.odata.org/V3/Northwind/Northwind.svc/Customers?$inlinecount=allpages&$top=5&$skip=10000